recipes-extended/xen/files/libxl-argo-xswait.patch

```diff
@@ -85,22 +85,37 @@ PATCHES
 +
 +    if (rc) {
 +        if (rc == ERROR_TIMEDOUT)
-+            LOGD(ERROR, guest_domid,
-+                 "%s: xenstore wait timed out", xswait->what);
++            LOGD(ERROR, guest_domid, "%s: xswait timed out", xswait->what);
++        else
++            LOGD(ERROR, guest_domid, "%s: xswait error (rc=%d)", xswait->what,
++                 rc);
 +        goto out;
 +    }
 +
-+    if (!p)
-+        LOGD(ERROR, guest_domid, "argo-firewall-ready was unable to be read");
-+    else if (!strcmp(p, "1"))
-+        LOGD(ERROR, guest_domid, "unexpected argo-firewall-ready value");
++    if (!p) {
++        LOGD(DEBUG, guest_domid, "argo-firewall-ready was unable to be read");
++        return;
++    }
++
++    if (strcmp(p, "1")) {
++        LOGD(WARN, guest_domid, "unexpected argo-firewall-ready value\"%s\"",
++             p);
++        return;
++    }
++
++    LOGD(DEBUG, guest_domid, "argo firewall is ready");
 +
-+ out:
 +    libxl__xswait_stop(gc, xswait);
 +    libxl__multidev_begin(ao, &sdss->multidev);
 +    sdss->multidev.callback = spawn_stub_launch_dm;
 +    libxl__add_disks(egc, ao, dm_domid, dm_config, &sdss->multidev);
-+    libxl__multidev_prepared(egc, &sdss->multidev, 0);
++    libxl__multidev_prepared(egc, &sdss->multidev, rc);
++
++    return;
++
++ out:
++    libxl__xswait_stop(gc, xswait);
++    qmp_proxy_spawn_outcome(egc, sdss, rc);
```

Comment:
The control flow of these changes confuses me a little. The early returns in the error cases mean that the `libxl__*` calls are never executed, which might mean we don't clean up properly in the case of an error? Also, the new `out:` seems more like an `err:` now as the success case never hits that block.

Reply:
Yes, it's a little confusing - took me a second to remember.  argo_xswait_cb() is called immediately on xenstore registration, so the `if (!p)` case is not an error - we need to keep waiting to give it a second chance.  For `if (strcmp(p, "1")) {`, I chose to not fail on invalid values and give it more time to come get set.  i.e. it'll timeout if it doesn't change to the expected value.  It shouldn't happen, but it isn't necessarily fatal which is why it just returns.  Since the xswait hasn't been stopped, return means just keep waiting.  This is what happens programming with libxl's async paradigm.

Reply:
libxl has 436 "out:" vs. 36 "err:", so I think I was mainly following surrounding code.  I can change it if you want.

Reply:
How about changing the log messages to:
"Unable to read argo-firewall-ready - keep waiting"
and
"Unexpected argo-firewall-ready value\"%s\" - keep waiting"
?

Reply:
I think those log additions would make things a lot clearer, +1. The label name is less important, but looking at it some more I'm not sure if that block is needed anymore. We only jump to it once, and that's during the proper error case. Since it's just two lines you might as well just call them in that `if` block directly.

Reply:
Yes, you are correct that the block could be moved.  However, I was trying to follow the libxl "style" using the out label.  With all the async callbacks in libxl, they try to funnel exiting to the next step at the end of a callback.  The other side of that is the "keep waiting" cases return return in the middle as this particular callback will be called again for success or failure.

So I'd leave it with the "libxl" style, but I can change it if you really want.

Reply:
Ack. If that's the paradigm then it is what it is.

Btw - for some reason I don't have a `Resolve` button under my comments. If you have the means you can feel free to resolve whatever comment chains are done at this point.